### PR TITLE
Add a base openstack code name as a source for add_source()

### PR DESCRIPTION
**Pull request by ajkavanagh**
This is a 'smart' option, where the openstack-origin=ussuri, for
example, passing 'ussuri' to the add_source() function.  This, on focal,
will be the equivalent of 'distro' and on bionic, will be
'cloud:bionic-ussuri'.  On groovy, victoria is distro, etc.

Also support a 'ussuri/proposed' to allow proposed to be carried between
distro versions.  This is to simplify the logic for setting openstack
versions so that the version wanted is the name rather than 'distro'.

### CODE REVIEW

File: charmhelpers/fetch/ubuntu.py [hunk 1]
```diff
@@ -574,6 +622,10 @@ def add_source(source, key=None, fail_invalid=False):
       with be used.  If staging is NOT used then the cloud archive [3] will be
       added, and the 'ubuntu-cloud-keyring' package will be added for the
       current distro.
+    '<openstack-version>': translate to cloud:<release> based on the current
+      distro version (i.e. for 'ussuri' this will either be 'bionic-ussuri' or
+      'distro'.
+    '<openstack-version/proposed': as above, but for proposed.
```
Comment (ChrisMacNaughton) [inline charmhelpers/fetch/ubuntu.py:628]: Should this say: `'<openstack-version>/proposed'`?

Comment (ajkavanagh) [reply charmhelpers/fetch/ubuntu.py:628]: > Should this say: `'<openstack-version>/proposed'`?

Ack, yes, it should.  Good catch, thank you.

File: charmhelpers/fetch/ubuntu.py [hunk 2]
```diff
@@ -738,6 +799,74 @@ def _verify_is_ubuntu_rel(release, os_release):
...
+
+    The spec given is, say, 'ussuri', but this could apply cloud:bionic-ussuri
+    or 'distro' depending on whether the ubuntu release is bionic or focal.
+
+    :param openstack_release: the OpenStack codename to determine the release
+        for.
+    :type openstack_release: str
+    :raises: SourceConfigError
+    """
+    # So it's distro, which means it's a no-op so end here.
```
Comment (AurelienLourot) [inline charmhelpers/fetch/ubuntu.py:813]: Should this comment be moved somewhere else? It's not necessarily "distro" at this point. I suspect this line was moved by mistake from somewhere else?

File: tests/fetch/test_fetch_ubuntu.py [hunk 3]
```diff
@@ -730,6 +730,98 @@ def test_add_distro_source(self, log):
...
+        mock_get_distrib_codename.return_value = 'bionic'
+        fetch.add_source('ussuri')
+        mock_add_cloud_pocket.assert_called_once_with("bionic-ussuri")
+
+    @patch('charmhelpers.fetch.ubuntu._add_cloud_pocket')
+    @patch('charmhelpers.fetch.ubuntu.get_distrib_codename')
+    def test_add_bare_openstack_impossible_version(
+            self, mock_get_distrib_codename, mock_add_cloud_pocket):
+        mock_get_distrib_codename.return_value = 'xenial'
+        try:
```
Comment (AurelienLourot) [inline tests/fetch/test_fetch_ubuntu.py:754]: nitpick, you just reinvented this wheel:

```
with self.assertRaises(fetch.SourceConfigError):
    fetch.add_source('ussuri')
mock_add_cloud_pocket.assert_not_called()
```

**Review by AurelienLourot — COMMENTED** [review]
LGTM in general, thanks! Just two nitpicks

File: charmhelpers/fetch/ubuntu.py (hunk 2 above)
Comment (ajkavanagh) [reply charmhelpers/fetch/ubuntu.py:813]: Ack; left in by mistake.

File: tests/fetch/test_fetch_ubuntu.py (hunk 3 above)
Comment (ajkavanagh) [reply tests/fetch/test_fetch_ubuntu.py:754]: If you look at the rest of the file, it's the pattern that already exists.  Also, assertRaises() isn't actually available in this class for the bizarre reason that it uses testtools where it doesn't exist.  One of those oddities of the library.

**Review by AurelienLourot — APPROVED** [review]

**Pull request merged**